chapter8/self-modifying-agent/test_evolution.py

```
@@ -62,6 +62,35 @@ def test_manifest_contains_change_contract_canary_and_rollback(self):
+        for source in ("\n", "# only a comment\n"):
+            checks = validate_candidate(source, self.trajectories)
+            self.assertFalse(all(checks.values()))
+            self.assertFalse(checks["failure_replay"])
+
+    def test_unsafe_candidate_is_not_executed_during_validation(self):
+        # The security scan must run before the candidate is exec()'d, so a
+        # source flagged unsafe never runs its module-level side effects.
+        import os
+        import tempfile
+
+        marker = tempfile.mktemp()
```

Ruff (0.16.0)

[error] Use of insecure and deprecated function (`mktemp`)

(S306)